README
NAME
    Test::Roo - Composable, reusable tests with roles and Moo

VERSION
    version 1.003

SYNOPSIS
    Define test behaviors and required fixtures in a role:

        # t/lib/ObjectCreation.pm

        package ObjectCreation;
        use Test::Roo::Role;    # loads Moo::Role and Test::More

        requires 'class';       # we need this fixture

        test 'object creation' => sub {
            my $self = shift;
            require_ok( $self->class );
            my $obj  = new_ok( $self->class );
        };

        1;

    Provide fixtures and run tests from the .t file:

        # t/test.t

        use Test::Roo; # loads Moo and Test::More
        use lib 't/lib';

        # provide the fixture
        has class => (
            is      => 'ro',
            default => sub { "Digest::MD5" },
        );

        # specify behaviors to test
        with 'ObjectCreation';

        # give our subtests a pretty label
        sub _build_description { "Testing " . shift->class }

        # run the test with default fixture
        run_me;

        # run the test with different fixture
        run_me( { class => "Digest::SHA1" } );

        done_testing;

    Result:

        $ prove -lv t
        t/test.t ..
                ok 1 - require Digest::MD5;
                ok 2 - The object isa Digest::MD5
                1..2
            ok 1 - object creation
            1..1
        ok 1 - Testing Digest::MD5
                ok 1 - require Digest::SHA1;
                ok 2 - The object isa Digest::SHA1
                1..2
            ok 1 - object creation
            1..1
        ok 2 - Testing Digest::SHA1
        1..2
        ok
        All tests successful.
        Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.06 cusr  0.00 csys =  0.09 CPU)
        Result: PASS

DESCRIPTION
    This module allows you to compose Test::More tests from roles. It is
    inspired by the excellent Test::Routine module, but uses Moo instead of
    Moose. This gives most of the benefits without the need for Moose as a
    test dependency.

    Test files are Moo classes. You can define any needed test fixtures as
    Moo attributes. You define tests as method modifiers -- similar in
    concept to "subtest" in Test::More, but your test method will be passed
    the test object for access to fixture attributes. You may compose any
    Moo::Role into your test to define attributes, require particular
    methods, or define tests.

    This means that you can isolate test *behaviors* into roles which
    require certain test *fixtures* in order to run. Your main test file
    will provide the fixtures and compose the roles to run. This makes it
    easy to reuse test behaviors.

    For example, if you are creating tests for Awesome::Module, you could
    create the test behaviors as Awesome::Module::Test::Role and distribute
    it with your module. If another distribution subclasses Awesome::Module,
    it can compose the Awesome::Module::Test::Role behavior for its own
    tests.

    No more copying and pasting tests from a super class! Superclasses
    define and share their tests. Subclasses provide their own fixtures and
    run the tests.

USAGE
    Importing Test::Roo also loads Moo (which gives you strictures with
    fatal warnings and other goodies) and makes the current package a
    subclass of Test::Roo::Class.

    Importing also loads Test::More. No test plan is used. The
    "done_testing" function must be used at the end of every test file. Any
    import arguments are passed through to Test::More's "import" method.

    See also Test::Roo::Role for test role usage.

  Creating fixtures
    You can create fixtures with normal Moo syntax. You can even make them
    lazy if you want:

        has fixture => (
            is => 'lazy'
        );

        sub _build_fixture { ... }

    This becomes really useful with Test::Roo::Role. A role could define the
    attribute and require the builder method to be provided by the main test
    class.

  Composing test roles
    You can use roles to define units of test behavior and then compose them
    into your test class using the "with" function. Test roles may define
    attributes, declare tests, require certain methods and anything else you
    can regularly do with roles.

        use Test::Roo;

        with 'MyTestRole1', 'MyTestRole2';

    See Test::Roo::Role and the Test::Roo::Cookbook for details and
    examples.

  Setup and teardown
    You can add method modifiers around the "setup" and "teardown" methods
    and these will be run before tests begin and after tests finish
    (respectively).

        before  setup     => sub { ... };

        after   teardown  => sub { ... };

    You can also add method modifiers around "each_test", which will be run
    before and after every individual test. You could use these to prepare
    or reset a fixture.

        has fixture => ( is => 'lazy, clearer => 1, predicate => 1 );

        after  each_test => sub { shift->clear_fixture };

    Roles may also modify "setup", "teardown", and "each_test", so the order
    that modifiers will be called will depend on when roles are composed. Be
    careful with "each_test", though, because the global effect may make
    composition more fragile.

    You can call test functions in modifiers. For example, you could confirm
    that something has been set up or cleaned up.

        before each_test => sub { ok( ! shift->has_fixture ) };

  Running tests
    The simplest way to use Test::Roo with a single .t file is to let the
    "main" package be the test class and call "run_me" in it:

        # t/test.t
        use Test::Roo; # loads Moo and Test::More

        has class => (
            is      => 'ro',
            default => sub { "Digest::MD5" },
        );

        test 'load class' => sub {
            my $self = shift;
            require_ok( $self->class );
        }

        run_me;
        done_testing;

    Calling "run_me(@args)" is equivalent to calling
    "__PACKAGE__->run_tests(@args)" and runs tests for the current package.

    You may specify an optional description or hash reference of constructor
    arguments to customize the test object:

        run_me( "load MD5" );
        run_me( { class => "Digest::MD5" } );
        run_me( "load MD5", { class => "Digest::MD5" } );

    See Test::Roo::Class for more about the "run_tests" method.

    Alternatively, you can create a separate package (in the test file or in
    a separate .pm file) and run tests explicitly on that class.

        # t/test.t
        package MyTest;
        use Test::Roo;

        use lib 't/lib';

        has class => (
            is       => 'ro',
            required => 1,
        );

        with 'MyTestRole';

        package main;
        use strictures;
        use Test::More;

        for my $c ( qw/Digest::MD5 Digest::SHA/ ) {
            MyTest->run_tests("Testing $c", { class => $c } );
        }

        done_testing;

EXPORTED FUNCTIONS
    Loading Test::Roo exports subroutines into the calling package to
    declare and run tests.

  test
        test $label => sub { ... };

    The "test" function adds a subtest. The code reference will be called
    with the test object as its only argument.

    Tests are run in the order declared, so the order of tests from roles
    will depend on when they are composed relative to other test
    declarations.

  top_test
        top_test $label => sub { ... };

    The "top_test" function adds a "top level" test. Works exactly like
    "test" except it will not start a subtest. This is especially useful in
    very simple testing situations where the extra subtest level is just
    noise.

    So for example the following test

        # t/test.t
        use Test::Roo;

        has class => (
            is       => 'ro',
            required => 1,
        );

        top_test basic => sub {
            my $self = shift;

            require_ok($self->class);
            isa_ok($self->class->new, $self->class);
        };

        for my $c ( qw/Digest::MD5 Digest::SHA/ ) {
            run_me("Testing $c", { class => $c } );
        }

        done_testing;

    produces the following TAP

        t/test.t ..
            ok 1 - require Digest::MD5;
            ok 2 - The object isa Digest::MD5
            1..2
        ok 1 - Testing Digest::MD5
            ok 1 - require Digest::SHA1;
            ok 2 - The object isa Digest::SHA1
            1..2
        ok 2 - Testing Digest::SHA1
        1..2
        ok
        All tests successful.
        Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.06 cusr  0.00 csys =  0.09 CPU)
        Result: PASS

  run_me
        run_me;
        run_me( $description );
        run_me( $init_args   );
        run_me( $description, $init_args );

    The "run_me" function calls the "run_tests" method on the current
    package and passes all arguments to that method. It takes a description
    and/or a hash reference of constructor arguments.

DIFFERENCES FROM TEST::ROUTINE
    While this module was inspired by Test::Routine, it is not a drop-in
    replacement. Here is an overview of major differences:

    *   Test::Roo uses Moo; Test::Routine uses Moose

    *   Loading Test::Roo makes the importing package a class; in
        Test::Routine it becomes a role

    *   Loading Test::Roo loads Test::More; Test::Routine does not

    *   In Test::Roo, "run_test" is a method; in Test::Routine it is a
        function and takes arguments in a different order

    *   In Test::Roo, all role composition must be explicit using "with"; in
        Test::Routine, the "run_tests" command can also compose roles

    *   In Test::Roo, test blocks become method modifiers hooked on an empty
        method; in Test::Routine, they become methods run via introspection

    *   In Test::Roo, setup and teardown are done by modifying "setup" and
        "teardown" methods; in Test::Routine they are done by modifying
        "run_test"

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/Test-Roo/issues>. You will be notified
    automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/Test-Roo>

      git clone https://github.com/dagolden/Test-Roo.git

AUTHOR
    David Golden <dagolden@cpan.org>

CONTRIBUTORS
    *   Arthur Axel 'fREW' Schmidt <frioux@gmail.com>

    *   Diab Jerius <djerius@gmail.com>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2013 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Test-Roo

1.003     2014-06-27 23:04:39-04:00 America/New_York

    [ADDED]

    - Test::Roo now additionally exports 'top_test' for
      more simple testing (Arthur Axel fREW Schmidt)

1.002     2013-09-25 14:03:55 America/New_York

    [FIXES]

    - Compile test could hang on Windows

    [PREREQS]

    - Dropped configure_requires for ExtUtils::MakeMaker to 6.17

1.001     2013-09-11 10:13:42 America/New_York

    [CHANGED]

    - modifiers on 'each_test' are now included in the subtest block so
      that any tests run in the modifiers are listed with other tests in
      the block

    [DOCUMENTED]

    - Enhanced SYNOPSIS

1.000     2013-03-08 18:59:20 America/New_York

    [INCOMPATIBLE CHANGES]

    - importing Test::Roo makes the current class a subclass of
      Test::Roo::Class; all methods moved there instead of being exported
    
    - Test::Roo now only exports 'test' and 'run_me'

    - done_testing is never called automatically

    [ADDED]

    - Test::Roo::Class has 'run_tests' method for creating and running
      tests; can be called as object or class method

    - Test::Roo::Class has lazy 'description' attribute for generating
      descriptions from test attributes

    - Added 'each_test' hook for global actions before/after every test

    [DOCUMENTED]

    - Includes Test::Roo::Cookbook with examples

    - Test::Roo docs explain differences from Test::Routine

0.002     2013-02-25 14:25:09 America/New_York

    [ADDED]

    - Refactored test run code into a new 'run_me' method to allow manual
      control over test runs, including multiple, parameterized runs.

0.001     2013-02-25 12:23:27 America/New_York

    - First release



lib/Test/Roo.pm
use 5.008001;
use strictures;

package Test::Roo;
# ABSTRACT: Composable, reusable tests with roles and Moo
our $VERSION = '1.003'; # VERSION

use Test::More 0.96 import => [qw/subtest/];

use Sub::Install;

sub import {
    my ( $class, @args ) = @_;
    my $caller = caller;
    for my $sub (qw/test top_test run_me/) {
        Sub::Install::install_sub( { into => $caller, code => $sub } );
    }
    strictures->import; # do this for Moo, since we load Moo in eval
    eval qq{
        package $caller;
        use Moo;
        extends 'Test::Roo::Class'
    };
    if (@args) {
        eval qq{ package $caller; use Test::More \@args };
    }
    else {
        eval qq{ package $caller; use Test::More };
    }
    die $@ if $@;
}

sub test {
    my ( $name, $code ) = @_;
    my $caller  = caller;
    my $subtest = sub {
        my $self = shift;
        subtest $name => sub { $self->each_test($code) }
    };
    eval qq{ package $caller; after _do_tests => \$subtest };
    die $@ if $@;
}

sub top_test {
    my ( $name, $code ) = @_;
    my $caller = caller;
    my $test = sub { shift->each_test($code) };
    eval qq{ package $caller; after _do_tests => \$test };
    die $@ if $@;
}

sub run_me {
    my $class = caller;
    $class->run_tests(@_);
}

1;


# vim: ts=4 sts=4 sw=4 et:

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Roo - Composable, reusable tests with roles and Moo

=head1 VERSION

version 1.003

=head1 SYNOPSIS

Define test behaviors and required fixtures in a role:

    # t/lib/ObjectCreation.pm

    package ObjectCreation;
    use Test::Roo::Role;    # loads Moo::Role and Test::More

    requires 'class';       # we need this fixture

    test 'object creation' => sub {
        my $self = shift;
        require_ok( $self->class );
        my $obj  = new_ok( $self->class );
    };

    1;

Provide fixtures and run tests from the .t file:

    # t/test.t

    use Test::Roo; # loads Moo and Test::More
    use lib 't/lib';

    # provide the fixture
    has class => (
        is      => 'ro',
        default => sub { "Digest::MD5" },
    );

    # specify behaviors to test
    with 'ObjectCreation';

    # give our subtests a pretty label
    sub _build_description { "Testing " . shift->class }

    # run the test with default fixture
    run_me;

    # run the test with different fixture
    run_me( { class => "Digest::SHA1" } );

    done_testing;

Result:

    $ prove -lv t
    t/test.t ..
            ok 1 - require Digest::MD5;
            ok 2 - The object isa Digest::MD5
            1..2
        ok 1 - object creation
        1..1
    ok 1 - Testing Digest::MD5
            ok 1 - require Digest::SHA1;
            ok 2 - The object isa Digest::SHA1
            1..2
        ok 1 - object creation
        1..1
    ok 2 - Testing Digest::SHA1
    1..2
    ok
    All tests successful.
    Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.06 cusr  0.00 csys =  0.09 CPU)
    Result: PASS

=head1 DESCRIPTION

This module allows you to compose L<Test::More> tests from roles.  It is
inspired by the excellent L<Test::Routine> module, but uses L<Moo> instead of
L<Moose>.  This gives most of the benefits without the need for L<Moose> as a
test dependency.

Test files are Moo classes.  You can define any needed test fixtures as Moo
attributes.  You define tests as method modifiers -- similar in concept to
C<subtest> in L<Test::More>, but your test method will be passed the test
object for access to fixture attributes.  You may compose any L<Moo::Role> into
your test to define attributes, require particular methods, or define tests.

This means that you can isolate test I<behaviors> into roles which require
certain test I<fixtures> in order to run.  Your main test file will provide the
fixtures and compose the roles to run.  This makes it easy to reuse test
behaviors.

For example, if you are creating tests for Awesome::Module, you could create
the test behaviors as Awesome::Module::Test::Role and distribute it with
your module.  If another distribution subclasses Awesome::Module, it can
compose the Awesome::Module::Test::Role behavior for its own tests.

No more copying and pasting tests from a super class!  Superclasses define and
share their tests.  Subclasses provide their own fixtures and run the tests.

=head1 USAGE

Importing L<Test::Roo> also loads L<Moo> (which gives you L<strictures> with
fatal warnings and other goodies) and makes the current package a subclass
of L<Test::Roo::Class>.

Importing also loads L<Test::More>.  No test plan is used.  The C<done_testing>
function must be used at the end of every test file.  Any import arguments are
passed through to Test::More's C<import> method.

See also L<Test::Roo::Role> for test role usage.

=head2 Creating fixtures

You can create fixtures with normal Moo syntax.  You can even make them lazy if
you want:

    has fixture => (
        is => 'lazy'
    );

    sub _build_fixture { ... }

This becomes really useful with L<Test::Roo::Role>.  A role could define the
attribute and require the builder method to be provided by the main test class.

=head2 Composing test roles

You can use roles to define units of test behavior and then compose them into
your test class using the C<with> function.  Test roles may define attributes,
declare tests, require certain methods and anything else you can regularly do
with roles.

    use Test::Roo;

    with 'MyTestRole1', 'MyTestRole2';

See L<Test::Roo::Role> and the L<Test::Roo::Cookbook> for details and
examples.

=head2 Setup and teardown

You can add method modifiers around the C<setup> and C<teardown> methods and
these will be run before tests begin and after tests finish (respectively).

    before  setup     => sub { ... };

    after   teardown  => sub { ... };

You can also add method modifiers around C<each_test>, which will be
run before and after B<every> individual test.  You could use these to
prepare or reset a fixture.

    has fixture => ( is => 'lazy, clearer => 1, predicate => 1 );

    after  each_test => sub { shift->clear_fixture };

Roles may also modify C<setup>, C<teardown>, and C<each_test>, so the order
that modifiers will be called will depend on when roles are composed.  Be
careful with C<each_test>, though, because the global effect may make
composition more fragile.

You can call test functions in modifiers. For example, you could
confirm that something has been set up or cleaned up.

    before each_test => sub { ok( ! shift->has_fixture ) };

=head2 Running tests

The simplest way to use L<Test::Roo> with a single F<.t> file is to let the
C<main> package be the test class and call C<run_me> in it:

    # t/test.t
    use Test::Roo; # loads Moo and Test::More

    has class => (
        is      => 'ro',
        default => sub { "Digest::MD5" },
    );

    test 'load class' => sub {
        my $self = shift;
        require_ok( $self->class );
    }

    run_me;
    done_testing;

Calling C<< run_me(@args) >> is equivalent to calling
C<< __PACKAGE__->run_tests(@args) >> and runs tests for the current package.

You may specify an optional description or hash reference of constructor
arguments to customize the test object:

    run_me( "load MD5" );
    run_me( { class => "Digest::MD5" } );
    run_me( "load MD5", { class => "Digest::MD5" } );

See L<Test::Roo::Class> for more about the C<run_tests> method.

Alternatively, you can create a separate package (in the test file or in a
separate F<.pm> file) and run tests explicitly on that class.

    # t/test.t
    package MyTest;
    use Test::Roo;

    use lib 't/lib';

    has class => (
        is       => 'ro',
        required => 1,
    );

    with 'MyTestRole';

    package main;
    use strictures;
    use Test::More;

    for my $c ( qw/Digest::MD5 Digest::SHA/ ) {
        MyTest->run_tests("Testing $c", { class => $c } );
    }

    done_testing;

=for Pod::Coverage add_methods_here

=head1 EXPORTED FUNCTIONS

Loading L<Test::Roo> exports subroutines into the calling package to declare
and run tests.

=head2 test

    test $label => sub { ... };

The C<test> function adds a subtest.  The code reference will be called with
the test object as its only argument.

Tests are run in the order declared, so the order of tests from roles will
depend on when they are composed relative to other test declarations.

=head2 top_test

    top_test $label => sub { ... };

The C<top_test> function adds a "top level" test.  Works exactly like L</test>
except it will not start a subtest.  This is especially useful in very simple
testing situations where the extra subtest level is just noise.

So for example the following test

    # t/test.t
    use Test::Roo;

    has class => (
        is       => 'ro',
        required => 1,
    );

    top_test basic => sub {
        my $self = shift;

        require_ok($self->class);
        isa_ok($self->class->new, $self->class);
    };

    for my $c ( qw/Digest::MD5 Digest::SHA/ ) {
        run_me("Testing $c", { class => $c } );
    }

    done_testing;

produces the following TAP

    t/test.t ..
        ok 1 - require Digest::MD5;
        ok 2 - The object isa Digest::MD5
        1..2
    ok 1 - Testing Digest::MD5
        ok 1 - require Digest::SHA1;
        ok 2 - The object isa Digest::SHA1
        1..2
    ok 2 - Testing Digest::SHA1
    1..2
    ok
    All tests successful.
    Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.06 cusr  0.00 csys =  0.09 CPU)
    Result: PASS

=head2 run_me

    run_me;
    run_me( $description );
    run_me( $init_args   );
    run_me( $description, $init_args );

The C<run_me> function calls the C<run_tests> method on the current package
and passes all arguments to that method.  It takes a description and/or
a hash reference of constructor arguments.

=head1 DIFFERENCES FROM TEST::ROUTINE

While this module was inspired by L<Test::Routine>, it is not a drop-in
replacement.  Here is an overview of major differences:

=over 4

=item *

L<Test::Roo> uses L<Moo>; L<Test::Routine> uses L<Moose>

=item *

Loading L<Test::Roo> makes the importing package a class; in L<Test::Routine> it becomes a role

=item *

Loading L<Test::Roo> loads L<Test::More>; L<Test::Routine> does not

=item *

In L<Test::Roo>, C<run_test> is a method; in L<Test::Routine> it is a function and takes arguments in a different order

=item *

In L<Test::Roo>, all role composition must be explicit using C<with>; in L<Test::Routine>, the C<run_tests> command can also compose roles

=item *

In L<Test::Roo>, test blocks become method modifiers hooked on an empty method; in L<Test::Routine>, they become methods run via introspection

=item *

In L<Test::Roo>, setup and teardown are done by modifying C<setup> and C<teardown> methods; in L<Test::Routine> they are done by modifying C<run_test>

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/Test-Roo/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/Test-Roo>

  git clone https://github.com/dagolden/Test-Roo.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 CONTRIBUTORS

=over 4

=item *

Arthur Axel 'fREW' Schmidt <frioux@gmail.com>

=item *

Diab Jerius <djerius@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut


lib/Test/Roo/Role.pm
use 5.008001;
use strictures;

package Test::Roo::Role;
# ABSTRACT: Composable role for Test::Roo
our $VERSION = '1.003'; # VERSION

use Test::Roo (); # no imports!
use Sub::Install;

sub import {
    my ( $class, @args ) = @_;
    my $caller = caller;
    Sub::Install::install_sub(
        { into => $caller, code => 'test', from => 'Test::Roo' } );
    strictures->import; # do this for Moo, since we load Moo in eval
    eval qq{
        package $caller;
        use Moo::Role;
    };
    if (@args) {
        eval qq{ package $caller; use Test::More \@args };
    }
    else {
        eval qq{ package $caller; use Test::More };
    }
    die $@ if $@;
}

1;


# vim: ts=4 sts=4 sw=4 et:

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Roo::Role - Composable role for Test::Roo

=head1 VERSION

version 1.003

=head1 SYNOPSIS

A testing role:

    # t/lib/MyTestRole.pm
    package MyTestRole;
    use Test::Roo::Role; # loads Moo::Role and Test::More

    requires 'class';

    test 'object creation' => sub {
        my $self = shift;
        require_ok( $self->class );
        my $obj  = new_ok( $self->class );
    };

    1;

=head1 DESCRIPTION

This module defines test behaviors as a L<Moo::Role>.

=head1 USAGE

Importing L<Test::Roo::Role> also loads L<Moo::Role> (which gives you
L<strictures> with fatal warnings and other goodies).

Importing also loads L<Test::More>.  Any import arguments are passed through to
Test::More's C<import> method.

=head2 Creating and requiring fixtures

You can create fixtures with normal Moo syntax.  You can even make them lazy
if you want and require the composing class to provide the builder:

    has fixture => (
        is => 'lazy'
    );

    requires '_build_fixture';

Because this is a L<Moo::Role>, you can require any method you like, not
just builders.

See L<Moo::Role> and L<Role::Tiny> for everything you can do with roles.

=head2 Setup and teardown

You can add method modifiers around the C<setup> and C<teardown> methods and
these will be run before tests begin and after tests finish (respectively).

    before  setup     => sub { ... };

    after   teardown  => sub { ... };

You can also add method modifiers around C<each_test>, which will be
run before and after B<every> individual test.  You could use these to
prepare or reset a fixture.

    has fixture => ( is => 'lazy, clearer => 1, predicate => 1 );

    after  each_test => sub { shift->clear_fixture };

Roles may also modify C<setup>, C<teardown>, and C<each_test>, so the order
that modifiers will be called will depend on when roles are composed.  Be
careful with C<each_test>, though, because the global effect may make
composition more fragile.

You can call test functions in modifiers. For example, you could
confirm that something has been set up or cleaned up.

    before each_test => sub { ok( ! shift->has_fixture ) };

=for Pod::Coverage method_names_here

=head1 EXPORTED FUNCTIONS

Loading L<Test::Roo::Role> exports a single subroutine into the calling package
to declare tests.

=head2 test

    test $label => sub { ... };

The C<test> function adds a subtest.  The code reference will be called with
the test object as its only argument.

Tests are run in the order declared, so the order of tests from roles will
depend on when they are composed relative to other test declarations.

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut


lib/Test/Roo/Class.pm
use 5.008001;
use strictures;

package Test::Roo::Class;
# ABSTRACT: Base class for Test::Roo test classes
our $VERSION = '1.003'; # VERSION

use Moo;
use MooX::Types::MooseLike::Base qw/Str/;
use Test::More 0.96 import => [qw/subtest/];

#--------------------------------------------------------------------------#
# attributes
#--------------------------------------------------------------------------#

#pod =attr description
#pod
#pod A description for a subtest block wrapping all tests by the object.  It is a
#pod 'lazy' attribute.  Test classes may implement their own C<_build_description>
#pod method to create a description from object attributes.  Otherwise, the default
#pod is "testing with CLASS".
#pod
#pod =cut

has description => (
    is      => 'rw',
    isa     => Str,
    lazy    => 1,
    builder => 1,
);

sub _build_description {
    my $class = ref $_[0];
    return "testing with $class";
}

#--------------------------------------------------------------------------#
# class or object methods
#--------------------------------------------------------------------------#

#pod =method run_tests
#pod
#pod     # as a class method
#pod     $class->run_tests();
#pod     $class->run_tests($description);
#pod     $class->run_tests($init_args);
#pod     $class->run_tests($description $init_args);
#pod
#pod     # as an object method
#pod     $self->run_tests();
#pod     $self->run_tests($description);
#pod
#pod If called as a class method, this creates a test object using an optional hash
#pod reference of initialization arguments.
#pod
#pod When called as an object method, or after an object has been generated, this
#pod method sets an optional description and runs tests.  It will call the C<setup>
#pod method (triggering any method modifiers), will run all tests (triggering any
#pod method modifiers on C<each_test>) and will call the C<teardown> method
#pod (triggering any method modifiers).
#pod
#pod If a description is provided, it will override any initialized or generated
#pod C<description> attribute.
#pod
#pod The setup, tests and teardown will be executed in a L<Test::More> subtest
#pod block.
#pod
#pod =cut

sub run_tests {
    my $self = shift;
    # get hashref from end of args
    # if any args are left, it must be description
    my ( $desc, $args );
    $args = pop if @_ && ref $_[-1] eq 'HASH';
    $desc = shift;

    # create an object if needed and possibly update description
    $self = $self->new( $args || {} )
      if !ref $self;
    $self->description($desc)
      if defined $desc;

    # execute tests wrapped in a subtest
    subtest $self->description => sub {
        $self->setup;
        $self->_do_tests;
        $self->teardown;
    };
}

#--------------------------------------------------------------------------#
# private methods and stubs
#--------------------------------------------------------------------------#

#pod =method setup
#pod
#pod This is an empty method used to anchor method modifiers.  It should not
#pod be overridden by subclasses.
#pod
#pod =cut

sub setup { }

#pod =method each_test
#pod
#pod This method wraps the code references set by the C<test> function
#pod from L<Test::Roo> or L<Test::Roo::Role> in a L<Test::More> subtest block.
#pod
#pod It may also be used to anchor modifiers that should run before or after
#pod each test block, though this can lead to brittle design as modifiers
#pod will globally affect every test block, including composed ones.
#pod
#pod =cut

sub each_test {
    my ( $self, $code ) = @_;
    $code->($self);
}

#pod =method teardown
#pod
#pod This is an empty method used to anchor method modifiers.  It should not
#pod be overridden by subclasses.
#pod
#pod =cut

sub teardown { }

# anchor for tests as method modifiers
sub _do_tests { }

1;


# vim: ts=4 sts=4 sw=4 et:

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Roo::Class - Base class for Test::Roo test classes

=head1 VERSION

version 1.003

=head1 DESCRIPTION

This module is the base class for L<Test::Roo> test classes.  It provides
methods to run tests and anchor modifiers.  Generally, you should not extend
this class yourself, but use L<Test::Roo> to do so instead.

=head1 ATTRIBUTES

=head2 description

A description for a subtest block wrapping all tests by the object.  It is a
'lazy' attribute.  Test classes may implement their own C<_build_description>
method to create a description from object attributes.  Otherwise, the default
is "testing with CLASS".

=head1 METHODS

=head2 run_tests

    # as a class method
    $class->run_tests();
    $class->run_tests($description);
    $class->run_tests($init_args);
    $class->run_tests($description $init_args);

    # as an object method
    $self->run_tests();
    $self->run_tests($description);

If called as a class method, this creates a test object using an optional hash
reference of initialization arguments.

When called as an object method, or after an object has been generated, this
method sets an optional description and runs tests.  It will call the C<setup>
method (triggering any method modifiers), will run all tests (triggering any
method modifiers on C<each_test>) and will call the C<teardown> method
(triggering any method modifiers).

If a description is provided, it will override any initialized or generated
C<description> attribute.

The setup, tests and teardown will be executed in a L<Test::More> subtest
block.

=head2 setup

This is an empty method used to anchor method modifiers.  It should not
be overridden by subclasses.

=head2 each_test

This method wraps the code references set by the C<test> function
from L<Test::Roo> or L<Test::Roo::Role> in a L<Test::More> subtest block.

It may also be used to anchor modifiers that should run before or after
each test block, though this can lead to brittle design as modifiers
will globally affect every test block, including composed ones.

=head2 teardown

This is an empty method used to anchor method modifiers.  It should not
be overridden by subclasses.

=for Pod::Coverage each_test setup teardown

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.17;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Composable, reusable tests with roles and Moo",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Test-Roo",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "Test::Roo",
  "PREREQ_PM" => {
    "Moo" => "1.000008",
    "MooX::Types::MooseLike::Base" => 0,
    "Sub::Install" => 0,
    "Test::More" => "0.96",
    "strictures" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "lib" => 0,
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Capture::Tiny" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => 0,
  "List::Util" => 0,
  "Moo" => "1.000008",
  "MooX::Types::MooseLike::Base" => 0,
  "Sub::Install" => 0,
  "Test::More" => "0.96",
  "lib" => 0,
  "strict" => 0,
  "strictures" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



